File: old/old_xml-and-json/src/test/java/org/tsdes/spring/xmlandjson/ConverterTest.java
package org.tsdes.spring.xmlandjson;

import org.junit.Test;
import org.tsdes.spring.xmlandjson.data.Post;
import org.tsdes.spring.xmlandjson.data.TopPosts;

import java.util.Date;

import static org.junit.Assert.*;

public class ConverterTest {


    @Test
    public void testXml(){

        Converter<TopPosts> converter = new ConverterImpl(TopPosts.class);
        TopPosts topPosts = getTopPosts();

        String xml = converter.toXML(topPosts);
        System.out.println(xml);

        TopPosts res = converter.fromXML(xml);

        verifyEquivalence(topPosts,res);
    }

    @Test
    public void testInvalidXml(){

        Converter<TopPosts> converter = new ConverterImpl(TopPosts.class,"/topposts.xsd");
        TopPosts topPosts = getTopPosts();

        String xml = converter.toXML(topPosts);
        assertNotNull(xml);

        topPosts.setRetrievedTime(null);//this is a required field

        xml = converter.toXML(topPosts); //should fail, as invalid
        assertNull(xml);
    }

    @Test
    public void testJSon(){

        Converter<TopPosts> converter = new ConverterImpl(TopPosts.class);

        TopPosts topPosts = getTopPosts();

        String json = converter.toJSon(topPosts);
        System.out.println(json);

        TopPosts res = converter.fromJSon(json);

        verifyEquivalence(topPosts,res);
    }

    @Test
    public void testToFromJSonXml(){

        Converter<TopPosts> converter = new ConverterImpl(TopPosts.class);
        TopPosts topPosts = getTopPosts();

        String json = converter.toJSon(topPosts);

        String xml = converter.fromJSonToXml(json);

        json = converter.fromXmlToJSon(xml);

        TopPosts res = converter.fromJSon(json);

        verifyEquivalence(topPosts,res);
    }


    @Test
    public void testSize(){

        Converter<TopPosts> converter = new ConverterImpl(TopPosts.class);
        TopPosts topPosts = getTopPosts();

        String json = converter.toJSon(topPosts);
        String xml  = converter.toXML(topPosts);

        //usually, XML is larger than JSON
        assertTrue(xml.length() > json.length());
    }

    @Test
    public void testAutomatedFiles(){

        Converter<org.tsdes.spring.xmlandjson.automated.TopPosts> automated =
                new ConverterImpl<>(org.tsdes.spring.xmlandjson.automated.TopPosts.class);

        org.tsdes.spring.xmlandjson.automated.TopPosts tp = new org.tsdes.spring.xmlandjson.automated.TopPosts();

        String xml = automated.toXML(tp);

        org.tsdes.spring.xmlandjson.automated.TopPosts res = automated.fromXML(xml);
        assertEquals(tp.getRetrievedTime(), res.getRetrievedTime());
    }


    @Test
    public void testGeneratedFilesXML(){

        Converter<TopPosts> manual = new ConverterImpl(TopPosts.class);
        Converter<org.tsdes.spring.xmlandjson.automated.TopPosts> automated =
                new ConverterImpl<>(org.tsdes.spring.xmlandjson.automated.TopPosts.class);

        TopPosts mTP = getTopPosts();
        String mxml = manual.toXML(mTP);

        org.tsdes.spring.xmlandjson.automated.TopPosts aTP = automated.fromXML(mxml);
        String axml = automated.toXML(aTP);

        TopPosts res = manual.fromXML(axml);

        verifyEquivalence(mTP,res);
    }


    @Test
    public void testGeneratedFilesJSon(){

        Converter<TopPosts> manual = new ConverterImpl(TopPosts.class);
        TopPosts topPosts = getTopPosts();

        String mxml = manual.toXML(topPosts);
        String mjson = manual.toJSon(topPosts);

        Converter<org.tsdes.spring.xmlandjson.automated.TopPosts> automated =
                new ConverterImpl<>(org.tsdes.spring.xmlandjson.automated.TopPosts.class);

        org.tsdes.spring.xmlandjson.automated.TopPosts aTP = automated.fromXML(mxml);
        String aXml = automated.toXML(aTP);
        String aJS = automated.toJSon(aTP);

        assertEquals(mxml, aXml);
        /*
            They might or might not be equal. In this particular case, they
            are not, as the Java classes have different fields
         */
        assertNotEquals(mjson, aJS);
    }


    @Test
    public void testCustomJson(){

        Converter<TopPosts> converter = new ConverterCustomJson<TopPosts>(TopPosts.class);
        TopPosts topPosts = getTopPosts();

        String custom = converter.toJSon(topPosts);
        System.out.println("Custom JSON:\n" + custom);

        TopPosts backWithGson = converter.fromJSon(custom);

        verifyEquivalence(topPosts, backWithGson);
    }



    private void verifyEquivalence(TopPosts a, TopPosts b) {

        assertNotNull(a);
        assertNotNull(b);
        assertEquals(a.getRetrievedTime().toString(),b.getRetrievedTime().toString());
        assertEquals(a.getPosts().size(), b.getPosts().size());

        for(int i=0; i<a.getPosts().size(); i++) {
            assertEquals(a.getPosts().get(i), b.getPosts().get(i));
        }
    }

    private TopPosts getTopPosts(){

        TopPosts topPosts = new TopPosts();
        topPosts.setRetrievedTime(new Date());

        Post a = new Post();
        a.setAuthor("First");
        a.setContent("A post");
        a.setVotes(2);
        topPosts.addPost(a);

        Post b = new Post();
        b.setContent("A second post with no attributes");
        topPosts.addPost(b);

        return topPosts;
    }
}

File: advanced/data-format/src/test/kotlin/org/tsdes/advanced/dataformat/ConverterTest.kt
package org.tsdes.advanced.dataformat

import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test
import org.tsdes.advanced.dataformat.data.Post
import org.tsdes.advanced.dataformat.data.TopPosts
import java.util.*
import javax.xml.bind.annotation.XmlAccessType
import javax.xml.bind.annotation.XmlAccessorType
import javax.xml.bind.annotation.XmlRootElement

class ConverterTest{

    @Test
    fun testXml() {

        val converter = ConverterImp(TopPosts::class.java)
        val topPosts = getTopPosts()

        val xml = converter.toXML(topPosts)
        println(xml)

        val res = converter.fromXML(xml)

        verifyEquivalence(topPosts, res)
    }


    @Test
    fun testJSon() {

        val converter = ConverterImp(TopPosts::class.java)

        val topPosts = getTopPosts()

        val json = converter.toJSon(topPosts)
        println(json)

        val res = converter.fromJSon(json)

        verifyEquivalence(topPosts, res)
    }

    @Test
    fun testToFromJSonXml() {

        val converter = ConverterImp(TopPosts::class.java)
        val topPosts = getTopPosts()

        var json = converter.toJSon(topPosts)

        val xml = converter.fromJSonToXml(json)

        json = converter.fromXmlToJSon(xml)

        val res = converter.fromJSon(json)

        verifyEquivalence(topPosts, res)
    }


    @Test
    fun testSize() {

        val converter = ConverterImp(TopPosts::class.java)
        val topPosts = getTopPosts()

        val json = converter.toJSon(topPosts)
        val xml = converter.toXML(topPosts)

        //usually, XML is larger than JSon
        assertTrue(xml.length > json.length)
    }



    @Test
    fun testCustomJson() {

        val converter = ConverterCustomJson(TopPosts::class.java)
        val topPosts = getTopPosts()

        val custom = converter.toJSon(topPosts)
        println("Custom JSON:\n$custom")

        val backWithGson = converter.fromJSon(custom)

        verifyEquivalence(topPosts, backWithGson)
    }


    @XmlRootElement
    @XmlAccessorType(XmlAccessType.FIELD)
    private data class X (
            var x: Long = 0
    )

    @Test
    fun testLong(){

        val x = X(Long.MAX_VALUE)
        val converter = ConverterImp(X::class.java)

        val toXml = converter.toXML(x)
        val fromXml = converter.fromXML(toXml)
        assertEquals(x, fromXml)

        val toJson = converter.toJSon(x)
        val fromJson = converter.fromJSon(toJson)
        assertEquals(x, fromJson)

        /*
            Note: max long would be 9223372036854775807 , ie, 2^63
            The JSON will contain the right value here, but technically it would be
            wrong, as JSON does not have 64bit signed longs, but rather just doubles.
            For example, a browser like Chrome would not be able to hold such value,
            and could "silently" round it to: 9223372036854776000.
            You can try by pasting 9223372036854775807 into the JavaScript console in
            Developer Tools.
         */
    }


    private fun verifyEquivalence(a: TopPosts, b: TopPosts) {

        assertNotNull(a)
        assertNotNull(b)
        assertEquals(a.retrievedTime.toString(), b.retrievedTime.toString())
        assertEquals(a.posts!!.size, b.posts!!.size)

        for (i in 0 until a.posts!!.size) {
            assertEquals(a.posts!![i], b.posts!![i])
        }
    }

    private fun getTopPosts(): TopPosts {

        val topPosts = TopPosts(Date())

        val a = Post(content = "A post", author = "First", votes = 2)
        topPosts.addPost(a)

        val b = Post(content = "A second post with no attributes")
        topPosts.addPost(b)

        return topPosts
    }
}